src/runtime/FinalizationRegistryObject.cpp

```diff
@@ -0,0 +1,105 @@
+    : gc()
+    , m_weakRefTarget(weakRefTarget)
+    , m_heldValue(heldValue)
+    , m_unregisterToken(unregisterToken)
+{
+}
+
+void* FinalizationRegistryObjectItem::operator new(size_t size)
+{
+    static bool typeInited = false;
+    static GC_descr descr;
+    if (!typeInited) {
+        GC_word obj_bitmap[GC_BITMAP_SIZE(FinalizationRegistryObjectItem)] = { 0 };
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(FinalizationRegistryObjectItem, m_weakRefTarget));
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(FinalizationRegistryObjectItem, m_heldValue));
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(FinalizationRegistryObjectItem, m_unregisterToken));
+        descr = GC_make_descriptor(obj_bitmap, GC_WORD_LEN(FinalizationRegistryObjectItem));
+        typeInited = true;
+    }
+    return GC_MALLOC_EXPLICITLY_TYPED(size, descr);
+}
+
+Value FinalizationRegistryObjectItem::getUnregisterToken()
+{
+    return m_unregisterToken;
+}
+
+FinalizationRegistryObject::FinalizationRegistryObject(ExecutionState& state, Value cleanupCallback, Object* realm)
+    : FinalizationRegistryObject(state, state.context()->globalObject()->finalizationRegistryPrototype(), cleanupCallback, realm)
+{
+}
+
+FinalizationRegistryObject::FinalizationRegistryObject(ExecutionState& state, Object* proto, Value cleanupCallback, Object* realm)
+    : Object(state, proto)
+    , m_cleanupCallback(cleanupCallback)
+    , m_realm(realm)
+{
+}
+
+void* FinalizationRegistryObject::operator new(size_t size)
+{
+    static bool typeInited = false;
+    static GC_descr descr;
+    if (!typeInited) {
+        GC_word obj_bitmap[GC_BITMAP_SIZE(FinalizationRegistryObject)] = { 0 };
+        Object::fillGCDescriptor(obj_bitmap);
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(FinalizationRegistryObject, m_cleanupCallback));
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(FinalizationRegistryObject, m_realm));
+        GC_set_bit(obj_bitmap, GC_WORD_OFFSET(FinalizationRegistryObject, m_cells));
+        descr = GC_make_descriptor(obj_bitmap, GC_WORD_LEN(FinalizationRegistryObject));
+        typeInited = true;
+    }
+    return GC_MALLOC_EXPLICITLY_TYPED(size, descr);
+}
+
+bool FinalizationRegistryObject::deleteOperation(ExecutionState& state)
+{
+    return true;
+}
+
+void FinalizationRegistryObject::setCell(ExecutionState& state, FinalizationRegistryObjectItem* cell)
+{
+    m_cells.pushBack(cell);
+}
+
+bool FinalizationRegistryObject::deleteCell(ExecutionState& state, Value unregisterToken)
+{
+    bool removed = false;
+    for (size_t i = 0; i < m_cells.size(); i++) {
+        if (!m_cells[i]->getUnregisterToken().isEmpty() && m_cells[i]->getUnregisterToken().equalsToByTheSameValueAlgorithm(state, unregisterToken)) {
+            m_cells[i] = new FinalizationRegistryObjectItem(state, Value(Value::EmptyValue), Value(Value::EmptyValue), Value(Value::EmptyValue));
+            removed = true;
```

Comment:
If there is no need to keep the empty value for delete operation, how about using Vector's `erase` function?